beacon-chain/blockchain/head.go

```diff
@@ -148,17 +149,23 @@ func (s *Service) saveHeadNoDB(ctx context.Context, b *ethpb.SignedBeaconBlock,
 			return errors.Wrap(err, "could not retrieve head state in DB")
 		}
 	} else {
-		headState, err = s.beaconDB.State(ctx, r)
-		if err != nil {
-			return errors.Wrap(err, "could not retrieve head state in DB")
+		s.initSyncStateLock.RLock()
+		cachedHeadState, ok := s.initSyncState[r]
+		if ok {
+			// only copy for states on epoch boundaries.
+			if helpers.IsEpochStart(cachedHeadState.Slot()) {
+				headState = cachedHeadState.Copy()
+			} else {
+				headState = cachedHeadState
+				delete(s.initSyncState, r)
+			}
 		}
+		s.initSyncStateLock.RUnlock()
 		if headState == nil {
-			s.initSyncStateLock.RLock()
-			cachedHeadState, ok := s.initSyncState[r]
-			if ok {
-				headState = cachedHeadState
+			headState, err = s.beaconDB.State(ctx, r)
```

Comment:
if `headState` is nil here, do you need to regen the state?

Reply:
we wouldn't be able to save the state then. The assumption when coming to this method, is that we have a valid state stored, so I don't think its required to put state gen over here.

Reply:
Makes sense